library/Backends/llvm/test/Test/Codegen.hs

```diff
@@ -157,8 +222,44 @@ llvmRecordNameShadowingTest = testCase "LLVM codegen test with record name shado
       shadowedTerm = PassTypes.ScopedRecordDeclM shadowedTy addTerm
       annotatedShadowedTerm = llvmAnnotatedTerm intTy shadowedTerm
       compiled =
-        fromRight (P.error "trivial test compilation failed") $
+        fromRight (P.error "name shadowing test compilation failed") $
           Compilation.termLLVMToModule annotatedShadowedTerm
   output <- runIntModule compiled
   let casted :: Int32 = fromIntegral output
   casted @?= fromIntegral (shadowedValLeft + (shadowingVal * shadowedValRight))
+
+trivialLLVMCodegenSumTest :: TestTree
+trivialLLVMCodegenSumTest = testCase "Trivial LLVM codegen test with sum types" $ do
+  let sumTypeName = "testSumType"
+      intVariantName = "testVariantInt"
+      stringVariantName = "testVariantString"
+      bits = 8
+      intTy = llvmIntType bits
+      stringTy = llvmStringType
+      testSumTy = (sumTypeName, [(intVariantName, intTy), (stringVariantName, stringTy)])
+      sumType = PassTypes.SumType sumTypeName
+      intValToAdd = 1
+      intValToAddTo = 2
+      intCase = llvmAddFunction bits intValToAdd
+      valIfString = 5
+      stringCase = llvmConstantFunction llvmStringType intTy $ Primitive.LitInt valIfString
+      intSumVal = llvmAnnotatedTerm intTy $ llvmIntVal intValToAddTo
+      stringSumVal = llvmAnnotatedTerm stringTy $ llvmStringVal "testString"
+      intSumTerm = PassTypes.VariantM (sumTypeName, intVariantName, intSumVal)
+      annotatedIntSumTerm = llvmAnnotatedTerm sumType intSumTerm
+      stringSumTerm = PassTypes.VariantM (sumTypeName, stringVariantName, stringSumVal)
+      annotatedStringSumTerm = llvmAnnotatedTerm sumType stringSumTerm
+      intCaseTerm = PassTypes.MatchM (sumTypeName, annotatedIntSumTerm, [intCase, stringCase])
+      annotatedIntCaseTerm = llvmAnnotatedTerm intTy intCaseTerm
+      stringCaseTerm = PassTypes.MatchM (sumTypeName, annotatedStringSumTerm, [intCase, stringCase])
+      annotatedStringCaseTerm = llvmAnnotatedTerm intTy stringCaseTerm
+      mulOp = annotatedBinOp ASTTypes.i8 Primitive.Mul
+      mulTerm = annotatedBinOpApp intTy mulOp annotatedIntCaseTerm annotatedStringCaseTerm
+      term = PassTypes.ScopedSumDeclM testSumTy mulTerm
+      annotatedTerm = llvmAnnotatedTerm intTy term
+      compiled = case Compilation.termLLVMToModule annotatedTerm of
+        Left err -> P.error $ "LLVM sum types test compilation failed: " <> show err
+        Right t -> t
```

Comment:
I feel like we could maybe abstract this out a bit. Hard to read such tests just reading down it. Not sure how completely to break it up to be more recongizable at a glance